<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Error Handling</title>
</head>
<body>

    <script>

        /*
            Error Handling

        -   Saat terjadi error di program JavaScript, kadang kita tidak ingin program kita berhenti
        -   Di JavaScript, kita bisa menangkap jika terjadi error
        -   Kita bisa menggunakan try catch statement untuk manangkap error
        -   Pada block try, kita akan mencoba mengakses kode program yang bisa menyebabkan error,
            dan jika terjadi error, block try akan berhenti dan otomatis masuk ke block catch
        -   Jika tidak terjadi error, block catch tidak akan dieksekusi
        */
        
        class MathUtil {
            static sum(...numbers) {

                if (numbers.length === 0) {
                    throw new Error("Total Parameter Harus Lebih Dari 0");
                }


                let total = 0;
                for (const number of numbers) {
                    total += number;
                }
                return total;
            }
        }

        //  Error Handling
        try {
            console.info("eko");
            console.info(MathUtil.sum());   // Terjadi Error
        } catch (error) {
            console.info(`Terjadi Error: ${error.message}`);
        } finally {
            console.info("Program Selesai");
        }

        /*
            Kata Kunci finally

        -   Kadang kita ingin melakukan sesuatu entah itu terjadi error ataupun tidak
        -   Dalam try catch, kita bisa manambahkan block finally
        -   Block finally ini akan selalu dieksekusi setelah try catch selesai, entah terjadi error
            atau tidak, block finally akan selalu dieksekusi
        */



        /*
            Try Finally

        -   Kata kunci finally juga bisa digunakan tanpa perlu menggunakan catch
        -   Biasanya ini digunakan dalam kasus tertentu
        */

        class Counter {
            #counterValue = 1;

            next() {
                try {
                    return this.#counterValue;
                } finally {
                    this.#counterValue++;
                }
            }
        }

        const counter = new Counter();
        console.info(counter.next());
        console.info(counter.next());
        console.info(counter.next());


    </script>
    
</body>
</html>